### DISCUSSION TITLE
How to get several blocks based on a sample block

### DISCUSSION THREAD

**Discussion by moimp3** [rawsrc/PhpEcho#20] [Q&A] [answered]
Hi!
I am starting with GitHub and PhpEcho and have some difficulties.
The goal of this question is to have some frames based on the same block [as showed here](http://proginet.eu/).
I try with this code but it doesn't work as wanted:

File home.php:
<?php declare(strict_types=1);
/** @var rawsrc\PhpEcho\PhpEcho $this */ // HOME PAGE

require_once DIR_ROOT.'vendor'.DIRECTORY_SEPARATOR.'autoload.php';

use rawsrc\PhpEcho\PhpEcho;

PhpEcho::setTemplateDirRoot(DIR_ROOT.'view'.DIRECTORY_SEPARATOR.'fr');

$sites = ['sirep', 'sirap', 'genealium', 'astro', /* 'omnes', 'Pierre Paris', 'Famille Paris', */];

$home				= new PhpEcho('layout/main.php');
$home['title']		= 'Portail Proginet';
$home['description']= 'Portail de présentation des applications Proginet';
$home['language']	= $language;

$body = new PhpEcho('block/home.php');
$body['header']		= new PhpEcho('block/header.php');
$body['language']	= $language;
$body['menu']		= new PhpEcho('block/menu.php');
$body['mainFrame']	= new PhpEcho('block/mainFrame.php');

foreach ( $sites as $key=>$value )
{
	switch ($value)
	{
		case 'sirep':
			$site = new PhpEcho('block/appli.php');
			$site['style']		= 'background-color:#b0cf77';
			$site['imgSrc']		= 'images/logoSirep.png';
			$site['goal']		= 'Pour les entreprises';
			$site['title']		= 'La prospection commerciale';
			$site['text']		= 
				"<ul>				
					<li>Assurez le suivi commercial de vos clients et prospects</li>
					<li>Gérez les rappels de vos clients</li>
					<li>Notez les résumés de vos contacts</li>
					<li>Fonctions spécifiques B to B (d'entreprise à entreprises)</li>
				</ul>";
			$site['link']		= 'http://sirep.proginet.local';
			$mainFrame[$value]	= $site;
			break;
		case 'sirap':
			$site = new PhpEcho('block/appli.php');
			$site['style']		= 'background-color:rgb(255,255,130)';
			$site['imgSrc']		= 'images/logoSirap.png';
			$site['goal']		= 'Pour les associations et les collectivités';
			$site['title']		= 'La gestion des compétences pour associations et collectivités';
			$site['text']		= 'Recensez et contactez vos bénévoles et vos salariés. Rapprochez leurs compétences de vos activités.';
			$site['link']		= 'http://sirap.proginet.local';
			$mainFrame[$value]	= $site;
			break;
		case 'genealium':
			$site = new PhpEcho('block/appli.php');
			$site['style']		= 'background-color:blue; color:white';
			$site['imgSrc']		= 'images/logoGenealium.png';
			$site['goal']		= 'Pour les particuliers ou les associations';
			$site['title']		= "Plus qu'une généalogie, l'histoire de toute votre famille !";
			$site['text']		= 'Une personne met les informations à jour. Toute la famille en bénéficie.';
			$site['link']		= 'http://genealium.proginet.local';
			$mainFrame[$value]	= $site;
			break;
	}
}

$body['mainFrame']	= $mainFrame;

$home['body']		= $body;

echo $home;

File mainFrame.php:
<article id="right">

<p class="center"><a href="benefits.php" class="benefitsButton vCenter">Vos avantages</a></p> 

<div class="content">
	<?php // tried lines
	/*
	<?= $this['content'] ?>
	<?= $this['sirep'] ?>
	*/ ?>
</div>

</article>

**Comment by rawsrc** [comment]
Salut moimp3, tu peux utiliser la balise code (CTRL+e) pour améliorer la lisibilité de tes messages.

Comment as-tu découpé ta page en blocs ?

**Comment by rawsrc** [comment]
Voici comment d'après ton site j'aurais codé un bloc  dans `/view/block/app.article.php`: 
```php
<?php /** @var rawsrc\PhpEcho\PhpEcho $this */ ?>
<article id="<?= $this['id'] ?>">>">
    <img src="<?= $this['logo.img'] ?>" alt="logo <?= $this['logo.name'] ?>" />
    <?= $this['goal'] ?>
    <h3><?= $this['title'] ?></h3>
    <ul>
    <?php foreach ($this['text'] as $text) { ?>
        <li><?= $text ?></li>
    <?php } ?>
    </ul>
    <p class="centrer"><a href="<?= $this['app.url'] ?>" class="bouton">Accéder</a></p>
</article>
```
Comme tu as besoin de 4 blocs sur ta page, tu devras faire 4 instances du même bloc en passant les valeurs attendues par bloc pour finaliser le rendu. Quelque chose comme ça : 
```php
new PhpEcho('block/app.article.php', [
    'id' => 'sirep',
    'logo.img' => 'images/logoSirep.png',
    'logo.name' => 'sirep',
    'goal' => 'Pour les entreprises',
    'title' => 'La prospection commerciale',
    'app.url' => '/appSirep/index.php',
    'text' => [
        'Assurez le suivi commercial de vos clients et prospects',
        'Gérez les rappels de vos clients', 
        'Notez les résumés de vos contacts',
        "Fonctions spécifiques B to B (d'entreprise à entreprises)",
    ],
]);
```
À toi de choisir si les infos proviennent d'une base de données, d'un fichier de paramètres ou de tes petits doigts...

Je te conseille aussi de créer autant de page basées sur `ViewBuilder` que nécessaire

à la prochaine

**rawsrc**

**Comment by moimp3** [comment]
C'est la toute première fois que j'utilise ce site et je n'en connais pas encore toutes les astuces. Par exemple, je n'ai pas trouvé comment avoir une coloration syntaxique.

La difficulté n'est pas d'écrire les blocs de chaque appli mais de les intégrer dans un bloc parent des 4 blocs.

Code de la page page/home.php
```
<?php declare(strict_types=1);
/** @var rawsrc\PhpEcho\PhpEcho $this */ // HOME PAGE

require_once DIR_ROOT.'vendor'.DIRECTORY_SEPARATOR.'autoload.php';

use rawsrc\PhpEcho\PhpEcho;

PhpEcho::setTemplateDirRoot(DIR_ROOT.'view'.DIRECTORY_SEPARATOR.'fr');

$sites = ['sirep', 'sirap', 'genealium', 'astro',];

$home			= new PhpEcho('layout/main.php');
$home['title']		= 'Portail Proginet';
$home['description']= 'Portail de présentation des applications Proginet';
$home['language']	= $language;

$body = new PhpEcho('block/home.php');
$body['header']		= new PhpEcho('block/header.php');
$body['language']	= $language;
$body['menu']		= new PhpEcho('block/menu.php');
$body['mainFrame']	= new PhpEcho('block/mainFrame.php');

foreach ( $sites as $key=>$value )
{
	$site = new PhpEcho('block/appli.php');
	switch ($value)
	{
		case 'sirep':
			$site['style']		= 'background-color:#b0cf77';
			$site['imgSrc']		= 'images/logoSirep.png';
			$site['goal']		= 'Pour les entreprises';
			$site['title']		= 'La prospection commerciale';
			$site['text']		= 
				"<ul>				
					<li>Assurez le suivi commercial de vos clients et prospects</li>
					<li>Gérez les rappels de vos clients</li>
					<li>Notez les résumés de vos contacts</li>
					<li>Fonctions spécifiques B to B (d'entreprise à entreprises)</li>
				</ul>";
			$site['link']		= 'http://sirep.proginet.local';
			break;
		case 'sirap':
			$site['style']		= 'background-color:rgb(255,255,130)';
			$site['imgSrc']		= 'images/logoSirap.png';
			$site['goal']		= 'Pour les associations et les collectivités';
			$site['title']		= 'La gestion des compétences pour associations et collectivités';
			$site['text']		= 'Recensez et contactez vos bénévoles et vos salariés. Rapprochez leurs compétences de vos activités.';
			$site['link']		= 'http://sirap.proginet.local';
			break;
		case 'genealium':
			$site['style']		= 'background-color:blue; color:white';
			$site['imgSrc']		= 'images/logoGenealium.png';
			$site['goal']		= 'Pour les particuliers ou les associations';
			$site['title']		= "Plus qu'une généalogie, l'histoire de toute votre famille !";
			$site['text']		= 'Une personne met les informations à jour. Toute la famille en bénéficie.';
			$site['link']		= 'http://genealium.proginet.local';
			break;
		case 'astro':
			$site['style']		= 'background-color:#ddd';
			$site['imgSrc']		= '';
			$site['goal']		= 'Pour usage personnel';
			$site['title']		= 'Double astrologie';
			$site['text']		= "La double astrologie permet de définir les caractères de la personne. Ce site reproduit un livre et dipose donc d'un accès protégé.";
			$site['link']		= 'http://astro.local';
			break;
	}
	$mainFrame[$value]	= $site;
}

$body['mainFrame']	= $mainFrame;

$home['body']		= $body;

echo $home;
```

Je te mets ici les différents blocs à incorporer dans cette page tels que je les écrits:
Bloc block/mainFrame.php qui contient l'ensemble des applis:
```
<article id="right">

<p class="center"><a href="benefits.php" class="benefitsButton vCenter">Vos avantages</a></p>

<div class="content">
	<?php // C'est ici que je dois intégrer les 4 blocs appli
</div>

</article>
```

Bloc /block/appli.php:
```
<article class="appli" style="<?= $this['style'] ?>">
	<img src="<?= $this['imgSrc'] ?>" alt="logo">
	<?= $this['goal'] ?>
	<h3><?= $this['title'] ?></h3>
	<?= $this->raw('text') ?>
	<p class="center"><a href="<?= $this['link'] ?>" class="button">Accéder</a></p>
</article>

```
Je ne parviens pas à incorporer les 4 blocs appli dans le bloc /block/mainFrame

**Comment by rawsrc** [comment, marked as answer]
Salut moimp,

relis bien ce que tu fais : 
ici : `$body['mainFrame'] = new PhpEcho('block/mainFrame.php');` tu définis une valeur (bloc PhpEcho) pour la clé `'mainFrame'`, puis un peu plus loin tu la redéfinis : `$body['mainFrame'] = $mainFrame;` avec `$mainFrame[$value] = $site;`

C'est normal que tu perdes l'affichage.

Plusieurs points importants : c'est plus qu'étrange de charger un autoload dans une vue... Une vue qui définit la racine des vues... Tout ça relève d'un fichier d'amorçage ou de configuartion (bootstrap)
L'idée d'utiliser le helper `->raw()` pour afficher du html... ce n'est pas une bonne idée du tout. C'est au bloc vue de se débrouiller pour afficher les données en provenance d'un tableau par exemple. Regarde mon exemple plus haut.

Dans tous les cas, voici comment tu aurais dû faire : 

```php
<?php /** @var rawsrc\PhpEcho\PhpEcho $this */ // HOME PAGE

require_once DIR_ROOT.'vendor'.DIRECTORY_SEPARATOR.'autoload.php';

use rawsrc\PhpEcho\PhpEcho;

PhpEcho::setTemplateDirRoot(DIR_ROOT.'view'.DIRECTORY_SEPARATOR.'fr');

$home = new PhpEcho('layout/main.php');
$home['title'] = 'Portail Proginet';
$home['description'] = 'Portail de présentation des applications Proginet';
$home['language'] = $language;

$body = new PhpEcho('block/home.php');
$body['header'] = new PhpEcho('block/header.php');
$body['language'] = $language;
$body['menu'] = new PhpEcho('block/menu.php');

$services = [];
$sites = ['sirep', 'sirap', 'genealium', 'astro'];

foreach ($sites as $value) {
    $block = new PhpEcho('block/appli.php');

    switch ($value) {
        case 'sirep':
            $block['style'] = 'background-color:#b0cf77';
            $block['imgSrc'] = 'images/logoSirep.png';
            $block['goal'] = 'Pour les entreprises';
            $block['title'] = 'La prospection commerciale';
            $block['link'] = 'http://sirep.proginet.local';
            $block['text'] = <<<'html'
              <ul>				
                  <li>Assurez le suivi commercial de vos clients et prospects</li>
                  <li>Gérez les rappels de vos clients</li>
                  <li>Notez les résumés de vos contacts</li>
                  <li>Fonctions spécifiques B to B (d'entreprise à entreprises)</li>
              </ul>               
            html;
        break;
        case 'sirap':
            $block['style'] = 'background-color:rgb(255,255,130)';
            $block['imgSrc'] = 'images/logoSirap.png';
            $block['goal'] = 'Pour les associations et les collectivités';
            $block['title'] = 'La gestion des compétences pour associations et collectivités';
            $block['link'] = 'http://sirap.proginet.local';
            $block['text'] = 'Recensez et contactez vos bénévoles et vos salariés. Rapprochez leurs compétences de vos activités.';
        break;
        case 'genealium':
            $block['style'] = 'background-color:blue; color:white';
            $block['imgSrc'] = 'images/logoGenealium.png';
            $block['goal'] = 'Pour les particuliers ou les associations';
            $block['title'] = "Plus qu'une généalogie, l'histoire de toute votre famille !";
            $block['link'] = 'http://genealium.proginet.local';
            $block['text'] = 'Une personne met les informations à jour. Toute la famille en bénéficie.';
        break;
        case 'astro':
            $block['style'] = 'background-color:#ddd';
            $block['imgSrc'] = '';
            $block['goal'] = 'Pour usage personnel';
            $block['title'] = 'Double astrologie';
            $block['link'] = 'http://astro.local';
            $block['text'] = "La double astrologie permet de définir les caractères de la personne. Ce site reproduit un livre et dispose donc d'un accès protégé.";
        break;
    }

    $services[]	= $block;
}

$body['mainFrame'] = new PhpEcho('block/mainFrame.php', ['services' => $services]);
$home['body'] = $body;

echo $home;
```
et le bloc vue mainFrame : 
```html
<article id="right">
<p class="center"><a href="benefits.php" class="benefitsButton vCenter">Vos avantages</a></p>
<div class="content">
    <?= $this['services'] ?>
</div>
</article>
```

**Comment by moimp3** [comment]
J'ai déménagé l'autoload dans le fichier index.
J'ai repris le code selon tes suggestions.
Maintenant tout fonctionne, donc un grand merci !!!
Il me reste juste un détail:
J'ai essayé de décomposer la ligne `$body['mainFrame'] = new PhpEcho('block/mainFrame.php', ['services' => $services]);` et je n'y suis pas parvenu.

**Comment by rawsrc** [comment]
salut,

tu peux passer directement à l'instance de PhpEcho, un tableau de valeurs clé => valeur utile pour finaliser le rendu. C'est exactement ce que je fais, je passe les valeurs (services) nécessaire au rendu du bloc. Je ne comprends pas ce que tu ne comprends pas.
